Blinding/src/DisplaceDigiTimes_module.cc

```diff
@@ -0,0 +1,211 @@
+    const auto pset = config().target_distribution.get<fhicl::ParameterSet>();
+    _target_distribution = BinnedSpectrum(pset);
+    _target_sampler = std::make_unique<CLHEP::RandGeneral>(_engine,
+                                               _target_distribution.getPDF(),
+                                               _target_distribution.getNbins());
+
+    // surfaces at which timing may be referenced
+    for (const auto& name: config().surface_ids()){
+      _surface_ids.emplace_back(name);
+    }
+
+    // reco
+    this->consumes<KalSeed>(_kalseed_tag);
+
+    // tracker
+    this->consumes<StrawDigiCollection>(_tracker_digi_tag);
+    this->consumes<StrawDigiADCWaveformCollection>(_tracker_digi_tag);
+    this->produces<StrawDigiCollection>();
+    this->produces<StrawDigiADCWaveformCollection>();
+
+    // calorimeter...
+  }
+
+  void DisplaceDigiTimes::produce(art::Event& event){
+    // tracker
+    auto digi_handle = event.getValidHandle<StrawDigiCollection>(_tracker_digi_tag);
+    auto adcs_handle = event.getValidHandle<StrawDigiADCWaveformCollection>(_tracker_digi_tag);
+    if (digi_handle->size() != adcs_handle->size()){
+      std::string msg = "mismatched StrawDigiCollection and StrawDigiWaveformCollection.";
+      msg += " sizes: " + std::to_string(digi_handle->size())
+           + " and " + std::to_string(adcs_handle->size());
+      throw cet::exception("DisplaceDigiTimes") << msg << std::endl;
+    }
+
+    // first, check that there is only a single KalSeed associated with
+    // these digis
+    auto kalseeds_handle = event.getValidHandle<KalSeedCollection>(_kalseed_tag);
+    if (kalseeds_handle->size() != 1){
+      std::string msg = "ambiguous KalSeedCollection (length " + std::to_string(kalseeds_handle->size()) + "), cannot assign reference time";
+      throw cet::exception("DisplaceDigiTimes") << msg << std::endl;
+    }
+    auto kalseed = (*kalseeds_handle)[0];
+
+    // search for downward-going KalIntersection with a matching surface
+    KalIntersection intersection;
+    auto sit = _surface_ids.begin();
+    bool adequate = false;
+    while ((!adequate) && (sit != _surface_ids.end())){
+      const auto& iits = kalseed.intersections(*sit);
+      // there are at most 2 valid intersections (upgoing / downgoing)
+      // so selecting on which one is downgoing is sufficient
+      for (const auto& iit: iits){
+        auto direction = iit->momentum3().Unit().Z();
+        if (0 < direction){
+          intersection = *iit;
+          adequate = true;
+        }
+      }
+      sit++;
+    }
+
+    // if no valid intersection found, then we cannot use this event
+    if (!adequate){
+      std::string msg = "no valid KalIntersection found";
+      throw cet::exception("DisplaceDigiTimes") << msg << std::endl;
+    }
+    // otherwise, take the time of the intersection as the reference time
+    double reference_time = intersection.time();
+    double target_time = this->sample_target_time();
```

Comment:
Does 'target_time' here refer to the physical stopping target? A KalSeed can intersect multiple foils, is the intent here to shift to a generic intersection, ignoring the specific foil? Or will the sampling time somehow take the foil into account?

Reply:
As I said in the out-of-thread comment, this was nomenclature clash (target of a mathematical operation vs stopping target). I think we use a location with a unique (at least, we expect unique) intersection. But we do have leeway in choosing where to sample, because the distribution we are shifting to is also within our control (and we can take it from data or MC as we see fit).